futures/src/lib.rs

```diff
@@ -1,164 +1,28 @@
-//!     // the races!
-//!     let data = tcp.and_then(|conn| download(conn));
-//!
-//!     // That all might take awhile, though, so let's not wait too long for it
-//!     // to all come back. The `select` combinator here returns a future which
-//!     // resolves to the first value that's ready plus the next future.
-//!     //
-//!     // Note we can also use the `then` combinator which is similar to
-//!     // `and_then` above except that it receives the result of the
-//!     // computation, not just the successful value.
-//!     //
-//!     // Again note that all the above calls to `and_then` and the below calls
-//!     // to `map` and such require no allocations. We only ever allocate once
-//!     // we hit the `Box::new()` call at the end here, which means we've built
-//!     // up a relatively involved computation with only one box, and even that
-//!     // was optional!
-//!
-//!     let data = data.map(Ok);
-//!     let timeout = timeout(timeout_dur).map(Err);
-//!
-//!     let ret = data.select(timeout).then(|result| {
-//!         match result {
-//!             Ok(result) => {
-//!                 let (data, _other_future) = result.split();
-//!                 match data {
-//!                     // One future succeeded, and it was the one which was
-//!                     // downloading data from the connection.
-//!                     Ok(data) => Ok(data),
-//!
-//!                     // The timeout fired, and otherwise no error was found, so
-//!                     // we translate this to an error.
-//!                     Err(_timeout) => {
-//!                         Err(io::Error::new(io::ErrorKind::Other, "timeout"))
-//!                     }
-//!                 }
-//!             }
-//!
-//!             // A normal I/O error happened, so we pass that on through.
-//!             Err(err) => Err(err.split().0),
-//!         }
-//!     });
-//!     return Box::new(ret);
-//!
-//!     fn resolve(url: &str) -> IoFuture<SocketAddr> {
-//!         // ...
-//! #       panic!("unimplemented");
-//!     }
-//!
-//!     fn connect(hostname: &SocketAddr) -> IoFuture<TcpStream> {
-//!         // ...
-//! #       panic!("unimplemented");
-//!     }
-//!
-//!     fn download(stream: TcpStream) -> IoFuture<Vec<u8>> {
-//!         // ...
-//! #       panic!("unimplemented");
-//!     }
-//!
-//!     fn timeout(stream: Duration) -> IoFuture<()> {
-//!         // ...
-//! #       panic!("unimplemented");
-//!     }
-//! }
-//! # fn main() {}
-//! ```
-//!
-//! Some more information can also be found in the [README] for now, but
-//! otherwise feel free to jump in to the docs below!
-//!
-//! [README]: https://github.com/alexcrichton/futures-rs#futures-rs
+//! **The best way to learn about this crate is through the [Asynchronous
+//! Programming in Rust](https://aturon.github.io/apr/) book**, which provides a
+//! comprehensive introduction. The docs within this crate, by contract, are
```

Comment:
"by contract" seems unnecessary

Reply:
😆 that was supposed to be "by *contrast*"

Reply:
lol :)